### PR TITLE
platforms get cached

### PR DESCRIPTION
**Pull request by rsherget**
#28442

### CODE REVIEW

**Comment by rsherget** [comment]
#build

File: dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/FeatureManager.java [hunk 1]
```diff
@@ -1694,15 +1696,19 @@ private boolean sameJavaSpecVersion() {
...
-                if (featureRepository.getFeature(resolvedFeature) == null) {
-                    return false;
+            && featureRepository.getConfiguredFeatures().equals(newConfiguredFeatures)){
+            if( !isBeta || 
+                (isBeta && featureRepository.getPlatforms().equals(newConfiguredPlatforms) 
+                && ((featureRepository.getPlatformEnvVar() == null && platformEnvironmentVariable == null) ||
+                (featureRepository.getPlatformEnvVar() != null && featureRepository.getPlatformEnvVar().equals(platformEnvironmentVariable))))) {
+                    
+                // check that all installed features are still installed
+                for (String resolvedFeature : featureRepository.getResolvedFeatures()) {
```
Comment (cbridgha) [inline dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/FeatureManager.java:1706]: This var name confuses me...   is this many features or one?

File: dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/FeatureManager.java [hunk 2]
```diff
@@ -1694,15 +1696,19 @@ private boolean sameJavaSpecVersion() {
...
-            for (String resolvedFeature : featureRepository.getResolvedFeatures()) {
-                if (featureRepository.getFeature(resolvedFeature) == null) {
-                    return false;
+            && featureRepository.getConfiguredFeatures().equals(newConfiguredFeatures)){
+            if( !isBeta || 
+                (isBeta && featureRepository.getPlatforms().equals(newConfiguredPlatforms) 
+                && ((featureRepository.getPlatformEnvVar() == null && platformEnvironmentVariable == null) ||
+                (featureRepository.getPlatformEnvVar() != null && featureRepository.getPlatformEnvVar().equals(platformEnvironmentVariable))))) {
+                    
+                // check that all installed features are still installed
```
Comment (cbridgha) [inline dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/FeatureManager.java:1705]: This block seems expensive?     Is this checking every feature installed in this local server?

File: dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/FeatureManager.java [hunk 3]
```diff
@@ -1694,15 +1696,19 @@ private boolean sameJavaSpecVersion() {
     private boolean areConfiguredFeaturesGood(Set<String> newConfiguredFeatures, Set<String> newConfiguredPlatforms) {
         if (!!!featureRepository.isDirty()
             && !!!featureRepository.hasConfigurationError()
-            && featureRepository.getConfiguredFeatures().equals(newConfiguredFeatures)
-            && featureRepository.getPlatforms().equals(newConfiguredPlatforms)) {
-            // check that all installed features are still installed
-            for (String resolvedFeature : featureRepository.getResolvedFeatures()) {
-                if (featureRepository.getFeature(resolvedFeature) == null) {
-                    return false;
+            && featureRepository.getConfiguredFeatures().equals(newConfiguredFeatures)){
```
Comment (cbridgha) [inline dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/FeatureManager.java:1700]: simplify...

File: dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/subsystem/FeatureRepository.java
Comment (tbitonti) [inline dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/subsystem/FeatureRepository.java:1]: Should we update FEATURE_CACHE_VERSION for the format which stores platform values?
If we switch between beta and non-beta in the same server, that will switch the reads and writes between the new and old formats.

Comment (tbitonti) [inline dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/subsystem/FeatureRepository.java:1]: Should 'setInstalledFeatures' pass in the current (on the repository) platforms and environment variable values?
That would ensure that those would be updated with 'no change'.

Comment (tbitonti) [inline dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/subsystem/FeatureRepository.java:1]: We might need 'isEmpty' checks in more places on the platform environment variable value.

**Comment by rsherget** [comment]
#build

**Comment by rsherget** [comment]
#build

**Comment by rsherget** [comment]
#libby

**Review by hlhoots — APPROVED** [review]
LGTM

**Review by cbridgha — APPROVED** [review]
Looks good

**Pull request merged**